ltmain_flang_darwin.diff needs changes to handle full path to the compiler

## Background information

### What version of Open MPI are you using? (e.g., v4.1.6, v5.0.1, git branch name and hash, etc.)

v5.0.10

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

via Spack

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

N/A

### Please describe the system on which you are running

* Operating system/version: MacOS Sequoia
* Computer hardware: M1
* Network type: unknown

-----------------------------

## Details of the problem

A spack installation of openmpi with Darwin-flang (`spack install -j4 openmpi %c,cxx,fortran=llvm`) fails with:

```
==> Error: ProcessError: Command exited with status 2:
    '/usr/local/esmf/spack/opt/spack/darwin-m1/gmake-4.4.1-ldxwjjipfnsfsjjwd6zg224iv253vmcz/bin/make' '-j4' 'V=1'

7 errors found in build log:
     16563    libtool: compile:  /usr/local/esmf/spack/opt/spack/darwin-m1/compiler-wrapper-1.0-4cpjroh23vzqehblxcil65kowvaulfzw/libexec/spack/clang/flang -I../../../../ompi/include -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmp
              i-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi/include -I../../../../ompi/mpi/fortran/use-mpi -I../../../.. -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-sr
              c -c mpi-ignore-tkr-sizeof.f90  -fPIC -o .libs/mpi-ignore-tkr-sizeof.o
     16564    libtool: compile:  /usr/local/esmf/spack/opt/spack/darwin-m1/compiler-wrapper-1.0-4cpjroh23vzqehblxcil65kowvaulfzw/libexec/spack/clang/flang -DHAVE_CONFIG_H -I. -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.1
              0-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi/mpi/fortran/use-mpi-ignore-tkr -I../../../../opal/include -I../../../../ompi/include -I../../../../oshmem/include -I../../../../ompi/mpiext/cuda/c -I../../../../ompi/mpiext/rocm/c -I../../../../ompi/includ
              e -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi/include -I../../../../ompi/mpi/fortran/use-mpi -I../../../.. -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg
              /T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src -c mpi-ignore-tkr.F90  -fPIC -o .libs/mpi-ignore-tkr.o
     16565    flang-22: warning: overriding deployment version from '10.16' to '11.0' [-Woverriding-deployment-version]
     16566    flang-22: warning: overriding deployment version from '10.16' to '11.0' [-Woverriding-deployment-version]
     16567    /bin/sh ../../../../libtool  --tag=FC   --mode=link /usr/local/esmf/spack/opt/spack/darwin-m1/compiler-wrapper-1.0-4cpjroh23vzqehblxcil65kowvaulfzw/libexec/spack/clang/flang -I../../../../ompi/include -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sac
              ks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi/include -I../../../../ompi/mpi/fortran/use-mpi -I../../../.. -I/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcen
              uag46i3zj73jd7wrsq5kp5vv/spack-src   -version-info 80:1:40 -Wl,-single_module -Wl,-flat_namespace -Wl,-commons,use_dylibs -L/usr/local/esmf/spack/opt/spack/darwin-m1/libevent-2.1.12-2xa7j3snlry22gqrc5hc6igcbnn6kz65/lib -L/usr/local/esmf/spack/opt/spack/da
              rwin-m1/hwloc-2.11.1-iabqrdaafg2wrsoptu3hx46b7ngv3k3t/lib -L/usr/local/esmf/spack/opt/spack/darwin-m1/pmix-6.1.0-y57qem7apnw3n55tdfvixjhhzy23k4vq/lib -o libmpi_usempi_ignore_tkr.la -rpath /usr/local/esmf/spack/opt/spack/darwin-m1/openmpi-5.0.10-rkb6wcenua
              g46i3zj73jd7wrsq5kp5vv/lib mpi-ignore-tkr.lo mpi-ignore-tkr-sizeof.lo ../../../../ompi/mpi/fortran/use-mpi/libusempi_internal_modules.la /private/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3z
              j73jd7wrsq5kp5vv/spack-src/opal/libopen-pal.la -levent_core -levent_pthreads -lhwloc -lpmix
     16568    libtool: link: /usr/local/esmf/spack/opt/spack/darwin-m1/compiler-wrapper-1.0-4cpjroh23vzqehblxcil65kowvaulfzw/libexec/spack/clang/flang --shared -Wl,-undefined -Wl,dynamic_lookup -o .libs/libmpi_usempi_ignore_tkr.40.dylib  .libs/mpi-ignore-tkr.o .libs/mp
              i-ignore-tkr-sizeof.o   -Wl,-force_load,../../../../ompi/mpi/fortran/use-mpi/.libs/libusempi_internal_modules.a  -L/usr/local/esmf/spack/opt/spack/darwin-m1/libevent-2.1.12-2xa7j3snlry22gqrc5hc6igcbnn6kz65/lib -L/usr/local/esmf/spack/opt/spack/darwin-m1/h
              wloc-2.11.1-iabqrdaafg2wrsoptu3hx46b7ngv3k3t/lib -L/usr/local/esmf/spack/opt/spack/darwin-m1/pmix-6.1.0-y57qem7apnw3n55tdfvixjhhzy23k4vq/lib /private/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag4
              6i3zj73jd7wrsq5kp5vv/spack-src/opal/.libs/libopen-pal.dylib -ldl -levent_core -levent_pthreads -lhwloc -lpmix  -Wl,-single_module -Wl,-flat_namespace -Wl,-commons -Wl,use_dylibs   -Wl,-install_name,`echo  /usr/local/esmf/spack/opt/spack/darwin-m1/openmpi-
              5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/lib | sed 's%^[ ]*%%'`/libmpi_usempi_ignore_tkr.40.dylib -compatibility_version 81 -current_version 81.1 -Wl,-single_module
  >> 16569    flang-22: error: unknown argument: '-compatibility_version'
  >> 16570    flang-22: error: unknown argument: '-current_version'
  >> 16571    flang-22: error: no such file or directory: '81'
  >> 16572    flang-22: error: no such file or directory: '81.1'
     16573    flang-22: warning: overriding deployment version from '10.16' to '11.0' [-Woverriding-deployment-version]
  >> 16574    make[2]: *** [Makefile:1586: libmpi_usempi_ignore_tkr.la] Error 1
     16575    make[2]: Leaving directory '/private/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi/mpi/fortran/use-mpi-ignore-tkr'
  >> 16576    make[1]: *** [Makefile:2711: all-recursive] Error 1
     16577    make[1]: Leaving directory '/private/var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-src/ompi'
  >> 16578    make: *** [Makefile:1540: all-recursive] Error 1

See build log for details:
  /var/folders/qt/mpzbvd5n2nn9x9g_jyv99344z_kvxg/T/sacks/spack-stage/spack-stage-openmpi-5.0.10-rkb6wcenuag46i3zj73jd7wrsq5kp5vv/spack-build-out.txt
```

The error appears to be the same as in this thread:

https://groups.google.com/a/lists.open-mpi.org/g/users/c/4mIxq9op75k

Specifically, the current patch in ltmain_flang_darwin.diff is set up to check the compiler name, but this check fails if the compiler is given as a full path rather than just `flang`.

I have proposed PRs to fix this, which I'll open into both main and v5.0.x. But I've run into trouble trying to test my changes directly, due to what seem to be differences in the spack build process when I'm trying to point to a branch rather than a release tarball. So for now, I'm doing a more kludgey workaround in spack, and I'll want help from someone to test my proposed changes to the ompi repository.

--- a/config/ltmain_flang_darwin.diff
+++ b/config/ltmain_flang_darwin.diff
@@ -1,13 +1,14 @@
 --- config/ltmain.sh
 +++ config/ltmain.sh
-@@ -9024,7 +9024,14 @@
+@@ -9024,7 +9024,15 @@
 	  compile_deplibs="$new_inherited_linker_flags $compile_deplibs"
 	  finalize_deplibs="$new_inherited_linker_flags $finalize_deplibs"
 	else
 -	  compiler_flags="$compiler_flags "`$ECHO " $new_inherited_linker_flags" | $SED 's% \([^ $]*\).ltframework% -framework \1%g'`
 +	  case $host in
 +	    *-*-darwin*)
-+	      case $CC in
++	      func_cc_basename "$CC"
++	      case $func_cc_basename_result in
 +		flang*) compiler_flags="$compiler_flags "`$ECHO " $new_inherited_linker_flags" | $SED 's% \([^ $]*\).ltframework% -Wl,-framework,\1%g'`;;
 +		*) compiler_flags="$compiler_flags "`$ECHO " $new_inherited_linker_flags" | $SED 's% \([^ $]*\).ltframework% -framework \1%g'`;;
 +	      esac;;
@@ -16,7 +17,7 @@
 	fi
        fi
        dependency_libs=$newdependency_libs
-@@ -9369,7 +9376,7 @@
+@@ -9369,7 +9377,7 @@
            # On Darwin other compilers
            func_cc_basename $CC
            case $func_cc_basename_result in
@@ -25,12 +26,13 @@
                    verstring="$wl-compatibility_version $wl$minor_current $wl-current_version $wl$minor_current.$revision"
                    ;;
                *)
-@@ -9869,7 +9876,10 @@
+@@ -9869,7 +9877,11 @@
        # Time to change all our "foo.ltframework" stuff back to "-framework foo"
        case $host in
 	  *-*-darwin*)
 -	  newdeplibs=`$ECHO " $newdeplibs" | $SED 's% \([^ $]*\).ltframework% -framework \1%g'`
-+	  case $CC in
++	  func_cc_basename "$CC"
++	  case $func_cc_basename_result in
 +	    flang*) newdeplibs=`$ECHO " $newdeplibs" | $SED 's% \([^ $]*\).ltframework% -Wl,-framework,\1%g'`;;
 +	    *) newdeplibs=`$ECHO " $newdeplibs" | $SED 's% \([^ $]*\).ltframework% -framework \1%g'`;;
 +	  esac

